### ISSUE TITLE
Memory leak when creating default storage node

### ISSUE THREAD

**Issue by ebrahimebrahim** [Slicer/Slicer#6099]
Creating a default storage node from the python console seems to cause a memory leak. Upon exit we get 

```txt
vtkDebugLeaks has detected LEAKS!
```

## Steps to reproduce

Launch Slicer, create a storable node, and then use the python console to call `CreateDefaultStorageNode` on it. Exit.

e.g.

```py
n = slicer.mrmlScene.AddNewNodeByClass("vtkMRMLMarkupsFiducialNode")
n.CreateDefaultStorageNode()
```

results in 

```txt
vtkDebugLeaks has detected LEAKS!
Class "vtkMRMLMarkupsJsonStorageNode" has 1 instance still around.
Class "vtkStringArray" has 2 instances still around.
Class "vtkObserverManager" has 1 instance still around.
Class "9vtkBufferIiE" has 1 instance still around.
Class "vtkIntArray" has 1 instance still around.
Class "vtkCommand or subclass" has 3 instances still around.
Class "vtkMRMLMessageCollection" has 1 instance still around.
```

and

```py
n = slicer.mrmlScene.AddNewNodeByClass("vtkMRMLScalarVolumeNode")
n.CreateDefaultStorageNode()
```

results in

```txt
vtkDebugLeaks has detected LEAKS!
Class "vtkStringArray" has 2 instances still around.
Class "vtkObserverManager" has 1 instance still around.
Class "9vtkBufferIiE" has 1 instance still around.
Class "vtkIntArray" has 1 instance still around.
Class "vtkCommand or subclass" has 3 instances still around.
Class "vtkMRMLVolumeArchetypeStorageNode" has 1 instance still around.
Class "vtkMRMLMessageCollection" has 1 instance still around.
```

## Environment
- Operating system: Ubuntu 20.04
- Slicer version: current master

This bug existed in 4.13.0-2021-12-06, and it did not exist in 4.11

**Comment by Connor-Bowley** [comment]
Most of the MRML nodes create the node using `vtkMRMLScene::CreateNodeByClass` which does not add the new object to the mrml scene (like `AddNewNodeByClass` does). You can prevent this vtkDebugLeak by doing the following

```
n = slicer.mrmlScene.AddNewNodeByClass("vtkMRMLScalarVolumeNode")
n.CreateDefaultStorageNode()
s = n.CreateDefaultStorageNode()
// use s
s.UnRegister(None)
```

**Comment by Connor-Bowley** [comment]
Note also that if you return a raw owning pointer from C++ to Python, you I not found a good way to delete the pointer. With VTK objects you can use UnRegister and if it is the last reference it will delete the memory, but if you have a non-VTK object, you can run into memory leaks.

**Comment by ebrahimebrahim** [comment]
Thanks @Connor-Bowley for the comments pointing me in the right direction!

I didn't have a proper understanding of the reference counting system for VTK objects. I understand this better now, but I still don't see _why_ `vtkMRMLScene::CreateNodeByClass` needs to be followed by `UnRegister`. Why would it create a node with an incorrect reference count? Is this really the intended behavior?

1. If it's not intended that `CreateNodeByClass` needs to be followed by `UnRegister`, then this is a bug and should be fixed.
2. If it's intended that `CreateNodeByClass` needs to be followed by `UnRegister`, then I'd like to understand why, but to close this issue I think we need the following:
  - The uses of `CreateDefaultStorageNode` in the [script repository](https://slicer.readthedocs.io/en/latest/developer_guide/script_repository.html) need to be followed by `UnRegister`, so that the snippets included there do not suggest memory-leaking code.
  - The API documentation for `vtkMRMLStorableNode::CreateDefaultStorageNode` (and the methods overriding this) as well as `vtkMRMLScene::CreateNodeByClass` should make it clear how to avoid the memory leak.

**Comment by jamesobutler** [comment]
The related Slicer issue about trying to get away from having to do UnRegister is https://github.com/Slicer/Slicer/issues/4205#issuecomment-604163727.

A past Slicer forum posts about the script repository having memory leaks. https://discourse.slicer.org/t/link-slices-views-by-default/1934/11?u=jamesobutler

There is documentation about the python memory management with factory methods, but I’m not sure if this documentation from the Wiki has been fully transferred over to the readthedocs yet. https://www.slicer.org/wiki/Documentation/Nightly/Developers/Tutorials/MemoryManagement#Python_scripts_and_scripted_modules

**Comment by ebrahimebrahim** [comment]
Thank you for finding those links @jamesobutler

My question about the reason for incorrect reference count is answered by the wiki page:
> Factory methods return an object that the caller owns (and thus the caller has to delete, with reference count >0) and Python adds an additional reference to this object when stored in a Python variable, resulting in a reference count of >1. To make sure that the object is deleted when the Python variable is deleted, we have to remove the additional reference that the factory method added. There is currently no automatic mechanism to remove that additional reference, so it has to be done manually by calling UnRegister

Closing because this issue can defer to #4205. It will be fixed one way or another when #4205 is resolved. I will add a comment over there about transferring the memory management documentation to readthedocs.
